### PR TITLE
Fix invalid DWARF for enums when using ThinLTO

### PR DESCRIPTION
**Pull request by philipc**
We were setting the same identifier for both the DW_TAG_structure_type
and the DW_TAG_variant_part. This becomes a problem when using ThinLTO
becauses it uses the identifier as a key for a map of types that is used
to delete duplicates based on the ODR, so one of them is deleted as a
duplicate, resulting in invalid DWARF.

The DW_TAG_variant_part isn't a standalone type, so it doesn't need
an identifier. Fix by omitting its identifier.

ODR uniquing is [enabled here](https://github.com/rust-lang/rust/blob/f21dee2c6179276321a88a63300dce74ff707e92/src/rustllvm/PassWrapper.cpp#L1101).

### CODE REVIEW

File: src/test/debuginfo/enum-thinlto.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// ignore-tidy-linelength
+
+// Require LLVM with DW_TAG_variant_part and a gdb that can read it.
+// min-system-llvm-version: 7.0
```
Comment (nikic) [inline src/test/debuginfo/enum-thinlto.rs]: To keep this in sync with #59377, this should check for 8.0.

**Comment by michaelwoerister** [comment]
Thanks for the PR, @philipc! I'll try to review shortly.

File: src/test/debuginfo/enum-thinlto.rs (hunk 1 above)
Comment (michaelwoerister) [reply src/test/debuginfo/enum-thinlto.rs]: Seconded.

**Comment by michaelwoerister** [comment]
This seems to make the infrastructure around composite types more complicated. Wouldn't it be easier to keep the ODR identifier for the `DW_TAG_variant_part` and just make it distinct from its parent's identifier (e.g. by appending `_variant_part` to it)?

**Comment by philipc** [comment]
Omitting the identifier seemed more correct to me, but yes that'll be simpler. I'll update the PR later.

**Comment by philipc** [comment]
I've given the variant parts their own identifier.

**Comment by michaelwoerister** [comment]
Thanks, @philipc! As far as I can tell these identifiers are solely used for LLVM uniquing and don't make it to actual debuginfo, so I don't think that giving an identifier to the variant part is especially wrong. 

That being said, this code is very old and was initially written against LLVM 3.x. It could probably do with a more fundamental overhaul. 

@bors r+

**Pull request merged**